### ISSUE TITLE
Difference applies color of removed object to interface

### ISSUE THREAD

**Issue by retospect** [openscad/openscad#501]
difference() {
  union() {
    translate([0,0,1]) color("green") cube([1,1,1]);
    color("red") cube([1,1,1]);
  };
  translate([-.5, -.5, -.5]) color("pink") cube([2,1,3]);
}

From two cubes of different color, a third, pink cube is removed. 
The exposed interface is pink (the color of the invisible object that was used to cut), instead of the color of the cubes that were there originally
##

**Comment by jonathanimb** [comment]
Workaround for some instances: make the cut transparent. 

	difference()
	{
		stack();
		color("Blue", 0) //0 is fully transparent
		cube([8,8,8], true);
	}

![stack_384px](https://cloud.githubusercontent.com/assets/8062300/5782322/b93016ac-9d7f-11e4-8827-5a74f0912884.png)

**Comment by atartanian** [comment]
both use cases are valid.

the current behaviour is useful if you want to see where the difference is being applied (usually on a single shape) which I would assume is the original use case.

obviously the currently requested function would also be useful especially when applied to assemblies.

could we just have a parameter on difference? something like 

```difference(cut_color = true) { }```

**Comment by atartanian** [comment]
or maybe only apply the color of the differenced object if it is explicitly given a color? so:

```
difference() {

    union(){

    translate([0,0,0])
    color("red")
    cylinder(h = 1, r = 20);
    
    translate([0,0,1])
    color("green")
    cylinder(h = 1, r = 20);
    
    translate([0,0,2])
    color("blue")
    cylinder(h = 1, r = 20);
    
    translate([0,0,3])
    color("yellow")
    cylinder(h = 1, r = 20);
    }

    translate([0,0,2.5])
    color([1,0,0])
    cube([20,20,20], center = true);
}```

**Comment by MichaelAtOz** [comment]
Sounds like a good solution.

**Comment by tcurdt** [comment]
Are there any plans on implementing this?

**Comment by kintel** [comment]
No plans, but the task is up for grabs if someone wants to take a stab at it.

**Comment by nophead** [comment]
I don't think it is possible with the way OpenCSG works. The second part of the difference will always be one colour I think because it is drawn as a single cube.

**Comment by kintel** [comment]
yep, this could easily turn into rewriting OpenCSG ...

**Comment by MethylBromide** [comment]
I run into the use case of cross-sectioning assemblies **all the time** and this is a huge PITA for me. If I could even just get the original color back with a call to **render** if the cutting object is "default color", that would be satisfactory. But separately sectioning each part in an assembly so I can re-color them afterwards, is just not going to fly.

**Comment by kintel** [comment]
This will still be hard to pull off for previews, but it should technically be doable for Manifold rendering with color support.
Question is how much we want preview and render to diverge.

**Comment by MethylBromide** [comment]
> Question is how much we want preview and render to diverge.
There are already lots of ways they differ, which is why I'm calling **render** in most cases -- this is by no means the only case where I need to. If that's what I have to do to make the color appear the way I need it to, that's what I'll do. But we're talking here about making something that many people reasonably want _easy_, versus nearly impossible.★

★ I did make it work on my current project by passing a cut object as a child to the module that makes the assembly view and having it individually difference each part before setting its color, but this is really a simple thing we want and we shouldn't have to twist our code into a pretzel and add many extra lines to make it behave reasonably. The module that creates an individual part should be able to set the part's color. ✦

✦ And no, the module for the specific part should not have to do the cutting -- the calling module has to translate it to the desired position and orientation befor the cut.

**Comment by jordanbrown0** [comment]
This question is key to color behavior. We should either keep it open and eventually change it, or explicitly close it as the permanent behavior.

**Comment by MichaelAtOz** [comment]
I think after you look at the winch case above, the cross section loosing colour is a significant issue. My vote Keep it open.

**Comment by jordanbrown0** [comment]
(Terminology:  by "volumetric color", I mean color that is throughout the volume of the object, versus "face coloring" that is only on the surface.  Existing OpenSCAD behavior is all based on face coloring.  Note also that volumetric color is closely related to handling multiple materials.)

I think that we need a new scheme for color handling - ref the mailing list thread on "volumetric color".  Volumetric color pretty clearly wants to maintain the color of an object even as you cut stuff away from it.

On the other hand, the current behavior is pretty deeply entrenched - it's been this way for >10 years, and people have known about it, and I'm pretty sure that at least some people have taken advantage of it.

Closing it as permanent behavior would just mean that we'd have to address volumetric coloring using new mechanisms that maintain the old face coloring mechanisms.

**Comment by MichaelAtOz** [comment]
I tried one of the suggestions above, difference() with transparent removed object.
Interesting results with Manifold.

See .scad in attached zip. [transparent-difference.zip](https://github.com/user-attachments/files/24989018/transparent-difference.zip)

color-difference.scad
F5

<img width="582" height="441" alt="Image" src="https://github.com/user-attachments/assets/be894384-f85a-42d3-a9ec-028dc8dbd546" />

F6 (Manifold)

<img width="498" height="399" alt="Image" src="https://github.com/user-attachments/assets/6cf7e6dc-75de-4434-b450-c8fafbd6e8c8" />

logo_test.scad
F6 (Manifold) [F5 not so good]

<img width="529" height="423" alt="Image" src="https://github.com/user-attachments/assets/872730ca-e3ce-4b20-83ed-24bc2dd05188" />

You get to see inside objects. That could be handy for debugging.

**Comment by kintel** [comment]
I don't think this is intentional; we should probably take a closer look at Manifold rendering and transparency..

**Comment by MichaelAtOz** [comment]
Could someone please give me a sanity check.

```
#cube(5);
//translate([-2.5,5.5,0]) cube(5);
translate([10,0,0]) #cube(5);
```
As above, I get two solid orange cubes (right of image). I expect transparent pink.
But remove the comment, I get two transparent pink cubes & a solid cube behind (left of image).
Both 2021.01 & 2026.01.25
Has there been a bug with `#` since 2021 that I've forgotten, or is there something wrong with my new-ish system?
<img width="1184" height="469" alt="Image" src="https://github.com/user-attachments/assets/dde965a6-76ec-4ed0-9ff8-162dc9b1f43e" />

**Comment by MichaelAtOz** [comment]
A new Nightly, 2026.02.01, has restored my expectation.

<img width="495" height="263" alt="Image" src="https://github.com/user-attachments/assets/2e1ef397-5aec-47b1-ad8f-ded735d4a033" />

But has that been a bug in 2021 that I don't remember??

**Comment by jordanbrown0** [comment]
[#5490 When the entire model is #ed, it previews as opaque orange instead of translucent pink](https://github.com/openscad/openscad/issues/5490)

**Comment by jordanbrown0** [comment]
As I noted there:
>The CSG Products list for the solo "#cube(10);" has the cube in both the normal and highlighted rendering chains. With a normal shape in the model the highlighted cube is only in the highlighted list.

**Comment by jordanbrown0** [comment]
2026.02.01 changed the behavior (and indeed made it better for that case), but still doesn't get all cases right.
```
difference() {
    #cube(10);
    sphere(10);
}
```
<img width="242" height="262" alt="Image" src="https://github.com/user-attachments/assets/a74fb648-f599-46d4-b6cd-1c63ab594e62" />

I would expect the entire cube to be visible, and the sphere to be invisible.

**Comment by jordanbrown0** [comment]
Let's take this # discussion over to #5490, which isn't quite on point but is less wrong than here.

**Comment by amatulic** [comment]
I would also like to see this issue fixed. I proposed in #4798 to have a new parameter in `difference()` to retain the color of the thing being cut, but that was closed as a duplicate of this issue.

**Comment by jordanbrown0** [comment]
As far as I understand, Preview is simply incapable of using the color of the positive object for newly created faces.  It doesn't *know* the color of the positive object.  (Preview difference is black magic.)

See the long discussions of volumetric color.  With volumetric color you would indeed need the color to be that of the positive object, but that doesn't mean we have a realistic way to do it (absent moving entirely to always rendering).

Simply "fixing" it would be a compatibility break.  There are models (notably at least one of mine) that depend on this behavior.  Having the change be opt-in, in some fashion, would address that concern.
